Patch AnswerHints

taniwallach opened the pull request:
Patch AnswerHints so that if a subroutine in the list "dies" there will still be a final result from the AnswerHints postfilter.
This is mean to address https://github.com/openwebwork/pg/issues/679.

drgrice1 commented on macros/answerHints.pl:
```diff
@@ -146,11 +146,16 @@ sub AnswerHints {
+	      # Make the call to run the function inside an eval to trap errors
+	      my $myResult = 0;
+	      eval { $myResult = &$wrong($correct,$student,$ans); };
+	      if ( $@ ) { $myResult = 0; }
+	      if ( $myResult ) {
```
The space inside the parentheses is inconsistent with the style of the rest of the file.  Those spaces are not preferable in any case according to the coding style that I like!

pstaabp replied:
Perhaps perltidy could be run on this file.  I figure eventually we'll run it on all files.

pstaabp replied:
Perhaps perltidy could be run on this file.  I figure eventually we'll run it on all files.

drgrice1 replied:
I agree.  I would like to see that as well.

taniwallach replied:
Dropped the space in the second commit.
perltidy applied in the third commit, but I needed to drop `--xci` and `--nlop` from `.perltidy` temporarily, as the version on my Debian machine does not support those options.

drgrice1 commented on macros/answerHints.pl:
```diff
@@ -146,11 +146,16 @@ sub AnswerHints {
+	      # Make the call to run the function inside an eval to trap errors
+	      my $myResult = 0;
+	      eval { $myResult = &$wrong($correct,$student,$ans); };
+	      if ( $@ ) { $myResult = 0; }
```
This line is not needed.  If the `$wrong` method has an error, then the assignment will not occur, and `$myResult` will still be `0` from the assignment on line 151.  Testing reveals that this is the case.

taniwallach replied:
Thanks. Dropped in commit 2 of the PR. However, I added a `or do` case which sets a fixed error message (in the results message box).

drgrice1 reviewed (approved):
I suppose this is okay.

I have mixed feelings about this though.

Your example code in the linked issue should set `$tmp = 0` and then wrap the `$tmp = $student->eval(x => 0)` in a perl `eval`.  Then the answerHint macro will work as expected and give the message that the condition requested in the problem is not satisfied.  This pull request will hide this error, and make it harder to see that the hints are not working in the intended way.

On the other hand, it doesn't hurt to catch errors from special cases don't matter that are accidentally missed by the author.

pstaabp reviewed (approved):
Fixes the error in the posted problem.  I would recommend the cleanups that @drgrice1 pointed out.

taniwallach commented:
@drgrice1 @pstaabp - Thanks. Perltidy changes are in the third commit.

> Your example code in the linked issue should set `$tmp = 0` and then wrap the `$tmp = $student->eval(x => 0)` in a perl `eval`. Then the answerHint macro will work as expected and give the message that the condition requested in the problem is not satisfied.

Thanks. I fixed the real problem (was more detailed, but the core error was the occurrence of a divide by zero error for a student answer at the initial condition point). I had not considered the need to such until this occurred.

I thought it would be helpful to document this in the Wiki, but the Wiki problems technique page about AnswerHints does not even mention the use of a subroutine test, so explaining there that evaluation at a test point could potentially lead to such an issue seems overboard there.

> This pull request will hide this error, and make it harder to see that the hints are not working in the intended way.

Good point. I added a fixed "error message" (as a result table message) for when the `eval` fails. That will make the existence of a problem visible to the end user (who may report it) but prevents the post filter from dying and messing up the allocation of weighed credit, etc.

I tried unsuccessfully to use the value of `$@` to retrieve the internal error message ("Divide by zero"). Using `warn ref($@);` could get that to show up as a warning message at the bottom, but I did not manage to get it out to put into the results table. It seems my understanding of what a failed `eval` is putting into `$@` is just not sufficient to dig it out. Ideas for how to do that are welcome.

drgrice1 reviewed (approved):
The perltidy result is a little different with the remove options added back in, but it will do for now.

I am not sure that I agree with displaying the "error during AnswerHints processing" as an answer message.  I think of those messages as informative messages for the student.  I think it might be better if that were a simple warn statement so that it appears with the information about reporting the issue to the professor and such.  This is really an issue with the problem.  In any case, this is will do for now also.

I am not sure why `$@` is not being set exactly.  It most likely has to do with the `$SIG{__WARN__}` handling inside the translator though.  The way that you dealt with it seems to be working well though.

pstaabp commented:
This looks good now.  I agree with @drgrice1 about the message.  We could say something along the lines of "An error occurred in this problem.  Please inform your instructor."

taniwallach commented:
> This looks good now. I agree with @drgrice1 about the message. We could say something along the lines of "An error occurred in this problem. Please inform your instructor."

Sounds good to me, but I would like to somehow mention AnswerHints in the message, to provide a hint to whoever debugs it of where to look. How about: "An error occurred in this problem. Please inform your instructor. (die in AnswerHints subroutine)"?

drgrice1 reviewed (changes requested):
I found an issue with the `eval` approach.  See my comment.

taniwallach commented on macros/answerHints.pl:
```diff
@@ -113,59 +113,81 @@ =head1 AnswerHints()
+						{
+							# Make the call to run the function inside an eval to trap errors
+							my $myResult = 0;
+							eval { $myResult = &$wrong($correct, $student, $ans); } or do {
```
@drgrice1 Thanks! That is a serious bug. Code patched as suggested and that fixes it.

taniwallach commented:
> I am not sure that I agree with displaying the "error during AnswerHints processing" as an answer message. I think of those messages as informative messages for the student. I think it might be better if that were a simple warn statement so that it appears with the information about reporting the issue to the professor and such.

Upon additional consideration, I changed the error message to a `warn` and the text to "An error occurred in this problem."

Since it is now a `warn` it will trigger the standard message above the problem: "Warning -- There may be something wrong with this question. Please inform your instructor including the warning messages below." Since the `warn` is from inside `answerHints.pl` the message below the problem provides an indication to the staff where to look for the bug.
